File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        boolean isFlag = false;
        for(String word: words){
            if(word.equals(target)){
                // target 단어가 words에 있으면
                isFlag = true;
                break;
            } 
        }
        
        if(!isFlag) return 0;
        else return bfs(begin, target, words);
    }
    
    static public int bfs(String begin, String target, String[] words){
        Queue<Node> queue = new LinkedList<>();
        // 단어랑 count값 저장
        queue.add(new Node(begin,0));
        
        boolean[] visited = new boolean[words.length];
        
        while(!queue.isEmpty()){
            Node now = queue.poll();
            if(now.word.equals(target)) return now.count;
            
            for (int i = 0; i < words.length; i++) {
                // 아직 방문 안 했고 + 글자가 딱 하나만 다르다면
                if (!visited[i] && canConvert(now.word, words[i])) {
                    visited[i] = true;
                    queue.add(new Node(words[i], now.count+1));
                }
            }
            
        }
        return 0;
    }
    
    static public boolean canConvert(String s1, String s2){
        int diff = 0;
        for(int i=0; i<s1.length();i++){
            if(s1.charAt(i) != s2.charAt(i)) diff +=1;
        }
        
        return diff == 1;
    }
}

class Node{
    String word;
    int count;
    
   public Node(String word, int count){
        this.word = word;
        this.count = count;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

struct Node{
    string word;
    int count;
};


bool canConvert(const string& s1, const string& s2){
    int diff = 0;
    for(int i=0; i<s1.size();i++){
        if(s1[i] != s2[i]) diff ++;
    }
    return diff == 1;
}


int bfs(string begin, string target, vector<string> words){
    queue<Node> queue;
    queue.push({begin, 0});
    
    vector<bool> visited(words.size(), false);
    
    while(!queue.empty()){
        Node now = queue.front();
        queue.pop();
        
        if(now.word == target) return now.count;
        for(int i=0; i<words.size();i++){
            if(!visited[i] && canConvert(words[i], now.word)){
                visited[i] = true;
                queue.push({words[i], now.count+1});
            }
        }
        
    }
    
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    
    bool isFlag = false;
    for(string word: words){
        if(word == target){
            isFlag = true;
            break;
        }
    }
    
    if(!isFlag) return 0;
    else return bfs(begin, target, words);
}
